relational_operators/HashJoinOperator.cpp

```diff
@@ -277,52 +377,59 @@ bool HashJoinOperator::getAllNonOuterJoinWorkOrders(
   if (blocking_dependencies_met_) {
     DCHECK(query_context != nullptr);
 
-    const Predicate *residual_predicate =
-        query_context->getPredicate(residual_predicate_index_);
-    const vector<unique_ptr<const Scalar>> &selection =
-        query_context->getScalarGroup(selection_index_);
-    InsertDestination *output_destination =
-        query_context->getInsertDestination(output_destination_index_);
-    const JoinHashTable &hash_table =
-        *(query_context->getJoinHashTable(hash_table_index_));
-
-    if (probe_relation_is_stored_) {
-      if (!started_) {
-        for (const block_id probe_block_id : probe_relation_block_ids_) {
-          container->addNormalWorkOrder(
-              new JoinWorkOrderClass(build_relation_,
-                                     probe_relation_,
-                                     join_key_attributes_,
-                                     any_join_key_attributes_nullable_,
-                                     probe_block_id,
-                                     residual_predicate,
-                                     selection,
-                                     hash_table,
-                                     output_destination,
-                                     storage_manager),
-              op_index_);
+  const Predicate *residual_predicate =
+      query_context->getPredicate(residual_predicate_index_);
+  const vector<unique_ptr<const Scalar>> &selection =
+      query_context->getScalarGroup(selection_index_);
+  InsertDestination *output_destination =
+      query_context->getInsertDestination(output_destination_index_);
+
+  if (probe_relation_is_stored_) {
+    if (!started_) {
+      if (probe_relation_.hasPartitionScheme()) {
+#ifdef QUICKSTEP_HAVE_LIBNUMA
+        if (probe_relation_.hasNUMAPlacementScheme()) {
+          addPartitionAwareWorkOrders<JoinWorkOrderClass>(container,
+                                                          query_context,
+                                                          storage_manager,
+                                                          residual_predicate,
+                                                          selection,
+                                                          output_destination);
         }
-        started_ = true;
+#endif
```

Comment:
Do we consider a case when the relation is partitioned but is not NUMA partitioned?